simd/src/Kokkos_SIMD_Common.hpp

```diff
@@ -46,6 +46,89 @@ inline constexpr simd_flags<simd_alignment_vector_aligned> simd_flag_aligned{};
 using element_aligned_tag = simd_flags<>;
 using vector_aligned_tag  = simd_flags<simd_alignment_vector_aligned>;
 
+namespace Impl {
+
+template <class BinaryOperation>
+struct is_basic_reduction_op {
+  static constexpr bool value =
+      std::is_same_v<BinaryOperation, std::plus<>> ||
+      std::is_same_v<BinaryOperation, std::multiplies<>> ||
+      std::is_same_v<BinaryOperation, std::bit_and<>> ||
+      std::is_same_v<BinaryOperation, std::bit_or<>> ||
+      std::is_same_v<BinaryOperation, std::bit_xor<>>;
+};
+
+template <class BinaryOperation>
+constexpr bool is_basic_reduction_op_v =
+    is_basic_reduction_op<BinaryOperation>::value;
+
+template <class T, class BinaryOperation>
+struct Identity {
+  KOKKOS_FORCEINLINE_FUNCTION
+  constexpr operator T() {
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+    if constexpr (std::is_same_v<BinaryOperation, std::plus<>>) {
+      return T();
+    } else if constexpr (std::is_same_v<BinaryOperation, std::multiplies<>>) {
+      return T(1);
+    } else if constexpr (std::is_same_v<BinaryOperation, std::bit_and<>>) {
+      return T(~T());
+    } else if constexpr (std::is_same_v<BinaryOperation, std::bit_or<>>) {
+      return T();
+    } else if constexpr (std::is_same_v<BinaryOperation, std::bit_xor<>>) {
+      return T();
+    } else {
+      return T();
+    }
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+  }
+};
```

Comment:
Any reason for moving all of this?

Reply:
OK. It threw me off a little and I started reviewing these lines not noticing that they aren't new.